app/clients.go

```diff
@@ -0,0 +1,128 @@
+package app
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/OAuth2withJWT/identity-provider/app/validation"
+)
+
+type ClientService struct {
+	repository ClientRepository
+}
+
+type Client struct {
+	Id          string
+	Name        string
+	Secret      string
+	Scope       string
+	RedirectURI string
+	CreatedBy   int
+}
+
+type CreateClientRequest struct {
+	Name        string
+	Scope       string
+	RedirectURI string
+	CreatedBy   int
+}
+
+type ClientCredentials struct {
+	Id     string
+	Secret string
+}
+
+func (req *CreateClientRequest) validateClientRegistrationFields(s *ClientService) error {
+	v := validation.New()
+	v.IsEmpty("Client Name", req.Name)
+	v.IsEmpty("Scope", req.Scope)
+	v.IsEmpty("Redirect URI", req.RedirectURI)
+	v.IsValidURI("Redirect URI", req.RedirectURI)
+	if s.hasClientWithName(req.Name) {
+		v.AddError("Client Name", fmt.Errorf("Client with that name already exists"))
+	}
+
+	if s.hasClientWithRedirectURI(req.RedirectURI) {
```

Comment:
Do we need to have unique Redirect URLs for Clients? What's the reasoning for this limitation?